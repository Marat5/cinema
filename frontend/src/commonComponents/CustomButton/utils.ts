import { COLORS } from '../../utils/constants';

export const getPulseLoaderColor = (look: 'primary' | 'secondary' | 'cancel') => {
  switch (look) {
    case 'primary':
      return COLORS.tertiaryColor;
    case 'secondary':
      return COLORS.primaryColor;
    case 'cancel':
      return COLORS.dangerColor;

    default:
      return COLORS.secondaryColor;
  }
};
