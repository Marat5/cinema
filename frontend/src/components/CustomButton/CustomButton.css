.CustomButton {
    width: 100%;
    border-width: 1px;
    border-radius: var(--field-border-radius);
    height: 36px;
    display: flex;
    align-items: center;
    justify-content: center;
}

.CustomButton:hover {
    cursor: pointer;
}

.CustomButton:disabled:hover {
    cursor: not-allowed;
}

.CustomButton_primary {
    background-color: var(--primary-color);
    color: var(--tertiary-color);
}

.CustomButton_primary:enabled:focus, .CustomButton_primary:enabled:hover {
    background-color: var(--primary-color-accent);
}

.CustomButton_secondary {
    color: var(--primary-color);
    border: 1px solid var(--primary-color);
}

.CustomButton_secondary:focus, .CustomButton_secondary:hover {
    color: var(--primary-color-accent);
    border-color: var(--primary-color-accent);
}

.CustomButton_cancel {
    color: var(--danger-color);
    border: 1px solid var(--danger-color);
}

.CustomButton_cancel:focus, .CustomButton_cancel:hover {
    background-color: var(--dark-grey-color);
}
