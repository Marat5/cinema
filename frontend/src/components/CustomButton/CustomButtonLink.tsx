import classNames from 'classnames';
import { Link } from 'react-router-dom';
import { BaseCustomButtonProps } from './CustomButton';

// CustomButton and CustomButtonLink are similar, have common props and use same styles
// But it is better for typing to separate them

type ButtonLinkProps = BaseCustomButtonProps & {
  to: string;
};

export const CustomButtonLink = ({
  text, look = 'primary', to, className
}: ButtonLinkProps) => (
  <Link
    to={to}
    className={classNames('CustomButton', `CustomButton_${look}`, className)}
  >
    {text}
  </Link>
);
