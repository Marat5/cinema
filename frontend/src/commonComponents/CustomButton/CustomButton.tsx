/* eslint-disable react/button-has-type */
import classNames from 'classnames';
import { MouseEvent } from 'react';
import { PulseLoader } from 'react-spinners';
import './CustomButton.css';
import { getPulseLoaderColor } from './utils';

// Shared common props for CustomButton and CustomButtonLink
export type BaseCustomButtonProps = {
  text: string
  look?: 'primary' | 'secondary' | 'cancel'
  className?: string
};

type Props = BaseCustomButtonProps & {
  type?: 'submit' | 'button' | 'reset',
  showLoadIndicator?: boolean
  onClick?: (e: MouseEvent<HTMLElement>) => void;
  disabled?: boolean
};

export const CustomButton = ({
  type = 'button', text, look = 'primary', showLoadIndicator, onClick, className, disabled
}: Props) => (
  <button
    onClick={onClick}
    disabled={showLoadIndicator || disabled}
    type={type}
    className={classNames('CustomButton', `CustomButton_${look}`, disabled && 'CustomButton_disabled', className)}
  >
    {showLoadIndicator ? <PulseLoader size={10} color={getPulseLoaderColor(look)} /> : text}
  </button>
);
