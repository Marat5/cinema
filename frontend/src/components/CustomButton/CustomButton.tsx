/* eslint-disable react/button-has-type */
import classNames from 'classnames';
import { MouseEvent } from 'react';
import { PulseLoader } from 'react-spinners';
import { COLORS } from '../../utils/constants';
import './CustomButton.css';

// Shared common props for CustomButton and CustomButtonLink
export type BaseCustomButtonProps = {
  text: string
  look?: 'primary' | 'secondary' | 'cancel'
};

type Props = BaseCustomButtonProps & {
  type: 'submit' | 'button' | 'reset',
  showLoadIndicator?: boolean
  onClick?: (e: MouseEvent<HTMLElement>) => void;
};

export const CustomButton = ({
  type, text, look = 'primary', showLoadIndicator, onClick
}: Props) => (
  <button
    onClick={onClick}
    disabled={showLoadIndicator}
    type={type}
    className={classNames('CustomButton', `CustomButton_${look}`)}
  >
    {showLoadIndicator ? <PulseLoader size={10} color={COLORS.tertiaryColor} /> : text}
  </button>
);
