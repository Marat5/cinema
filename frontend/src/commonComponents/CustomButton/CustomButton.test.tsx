import { fireEvent, render } from '@testing-library/react';
import { CustomButton } from './CustomButton';

const BUTTON_TEXT = 'Custom Button';

test('Renders custom button', () => {
  const onClick = jest.fn();

  const { asFragment, getByText } = render(<CustomButton
    onClick={onClick}
    text={BUTTON_TEXT}
  />);

  expect(asFragment()).toMatchSnapshot();
  expect(getByText(BUTTON_TEXT)).toBeInTheDocument();

  fireEvent.click(getByText(BUTTON_TEXT));
  expect(onClick).toBeCalledTimes(1);
});
