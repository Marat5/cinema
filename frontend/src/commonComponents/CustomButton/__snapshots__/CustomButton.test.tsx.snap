// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders custom button 1`] = `
<DocumentFragment>
  <button
    class="CustomButton CustomButton_primary"
    type="button"
  >
    Custom Button
  </button>
</DocumentFragment>
`;
